### PR TITLE
Mark tasks pending before submitting to coordinator

### PR DESCRIPTION
**Pull request by aherrmann**
Now, the executor can assume that any task it is given is already marked
as pending and that this is the only executor executing that task.
The interpreter is now responsible for checking if the item is already
complete and marking it as pending if not.

Previously, a long lasting coordinator, like Redis or SQLite, stored a
completed task for a long time. If the content store item was removed in
the meantime, the coordinator would refuse to resubmit the task to
reconstruct that item.

### CODE REVIEW

File: funflow/src/Control/FunFlow/ContentStore.hs [hunk 1]
```diff
@@ -395,17 +406,36 @@ waitUntilComplete store hash = lookupOrWait store hash >>= \case
...
+  Pending a -> wait a >>= \case
+    Completed item -> return $ Complete item
+    -- XXX: Consider extending 'Status' with a 'Failed' constructor.
+    --   If the store contains metadata as well, it could keep track of the
+    --   number of failed attempts and further details about the failure.
+    --   If an external task is responsible for the failure, the client could
+    --   choose to resubmit a certain number of times.
+    Failed -> throwIO $ FailedToConstruct hash
+  Complete item -> return $ Complete item
+  Missing dir -> return $ Missing dir
```
Comment (nc6) [inline funflow/src/Control/FunFlow/ContentStore.hs]: The semantics of this seem slightly confusing - e.g. the directory is no longer missing, since we have created it at https://github.com/tweag/funflow/pull/42/files#diff-06fe0eceda2b774b025d64b91b0c4f0cR417. Possibly expand the comments to detail what the various return cases mean?

File: funflow/src/Control/FunFlow/Exec/Simple.hs [hunk 2]
```diff
@@ -94,22 +86,18 @@ runFlowEx _ cfg store runWrapped confIdent flow input = do
...
-      CS.waitUntilComplete store chash >>= \case
-        Nothing -> fail "Remote process failed to construct item"
-        Just item -> return item
+      CS.constructOrWait store chash >>= \case
+        CS.Pending void -> absurd void
+        CS.Complete item -> return item
+        CS.Missing _ -> do
+          submitTask po $ TaskDescription chash (toTask x)
+          CS.waitUntilComplete store chash >>= \case
+            Nothing -> fail "Remote process failed to construct item"
```
Comment (nc6) [inline funflow/src/Control/FunFlow/Exec/Simple.hs]: Unrelated, but we shouldn't use `fail` here.

File: funflow/src/Control/FunFlow/External/Coordinator.hs [hunk 3]
```diff
@@ -51,8 +51,7 @@ class Coordinator c where
   initialise :: MonadIO m => Config c -> m (Hook c)
 
   -- | Submit a task to the task queue.
-  --   If this task is already known to the system and not completed, then
```
Comment (nc6) [inline funflow/src/Control/FunFlow/External/Coordinator.hs]: What's the motivation for this change? As I understand it, the executor should no longer store a 'completed' status, but I don't see why this would mean we'd want to restart 'in-progress' tasks?

File: funflow/src/Control/FunFlow/External/Coordinator/Redis.hs [hunk 4]
```diff
@@ -35,8 +35,8 @@ instance Coordinator Redis where
   -- | Create a redis connection
   initialise = liftIO . R.connect
 
-  submitTask conn td = unlessM (isInProgress conn $ td ^. tdOutput) $
```
Comment (nc6) [inline funflow/src/Control/FunFlow/External/Coordinator/Redis.hs]: Same question as above.

File: funflow/src/Control/FunFlow/External/Coordinator.hs (hunk 3 above)
Comment (aherrmann) [reply funflow/src/Control/FunFlow/External/Coordinator.hs]: The flow interpreter should first query the store if the item is missing/pending/complete,
and it should only submit a task to the coordinator if the item is missing. If the store considers the item missing, but the coordinator has a running task for it then something must have gone wrong. In particular, that means that the build directory into which the task would store it's output no longer exists and that that task is going to fail once it tries to write its output. Ideally we would kill the running task before submitting the new task. But, I'm not sure that we can do that in the general case, e.g. when the task is being executed on a different machine.

In any case, if this happens, then the system is in an invalid state.
Maybe it's safer to throw an exception in this case, instead of quietly resubmitting?

File: funflow/src/Control/FunFlow/ContentStore.hs (hunk 1 above)
Comment (aherrmann) [reply funflow/src/Control/FunFlow/ContentStore.hs]: I've added some explanation in the haddocks.

File: funflow/src/Control/FunFlow/Exec/Simple.hs (hunk 2 above)
Comment (aherrmann) [reply funflow/src/Control/FunFlow/Exec/Simple.hs]: I've added a corresponding exception type and use `throwM` instead of `fail` now.

**Comment by nc6** [comment]
Thanks for explanations, LGTM!
